### PR TITLE
Fix #61: Generate JsonProperty attribute for properties with defaults

### LINKED ISSUE #61: Emit a JsonProperty attribute on all non-required properties
Regardless of the presence of the `System.ComponentModel.DefaultValue` attribute (which we now emit for any property for which the schema specifies a default value; see #58), that default is ignored on deserialization unless the property is generated with `JsonProperty` attribute that specifies "populate default on deserialize".

We need to emit that attribute.

An `enhancement` if you're ok with the workaround of adding a CodeGenHint; otherwise a `bug`.

### PR DESCRIPTION
**Pull request by lgolding**

### CODE REVIEW

File: src/Json.Schema.ToDotNet.UnitTests/Assert.cs [hunk 1]
```diff
@@ -10,10 +10,13 @@ internal static class Assert
     {
         internal static void FileContentsMatchExpectedContents(
             TestFileSystem testFileSystem,
-            IDictionary<string, ExpectedContents> expectedContentsDictionary)
+            IDictionary<string, ExpectedContents> expectedContentsDictionary,
+            bool generateEqualityComparers)
```
Comment (lgolding) [inline src/Json.Schema.ToDotNet.UnitTests/Assert.cs]: >generateEqualityComparers [](start = 17, length = 25)

Before this, every unit test generated an equality comparer in addition to the class, and this assertion method relied on that. The test I just added has no implications for the equality comparer, so I wanted to make it possible to write a test that just generated the class. All the other changes, except for the one new test and the actual implementation, flow from this change.

File: src/Json.Schema.ToDotNet/SyntaxNodeExtensions.cs [hunk 2]
```diff
@@ -120,20 +120,17 @@ private static BaseTypeDeclarationSyntax AddGeneratedCodeAttribute(BaseTypeDecla
             AttributeListSyntax attributeListForGeneratedCodeAttribute =
                 SyntaxFactory.AttributeList(SyntaxFactory.SingletonSeparatedList(MakeGeneratedCodeAttribute()));
 
-            var enumDecl = typeDecl as EnumDeclarationSyntax;
-            if (enumDecl != null)
+            if (typeDecl is EnumDeclarationSyntax enumDecl)
```
Comment (lgolding) [inline src/Json.Schema.ToDotNet/SyntaxNodeExtensions.cs]: >enumDecl [](start = 50, length = 8)

Hygiene: Use C# 7 "pattern matching".

**Comment by lgolding** [comment]
>         protected override AttributeSyntax[] GeneratePropertyAttributes(string propertyName, string serializedName, bool isRequired, object defaultValue)

This method is too long and I should refactor it into helpers like `GenerateDataMemberAttribute`, `GenerateDefaultValueAttribute`, `GenerateJsonPropertyAttribute`...

---
Refers to: src/Json.Schema.ToDotNet/ClassGenerator.cs:252 in ad3e739. [](commit_id = ad3e7397e99d050f0a65872564a2063bb1d8f97d, deletion_comment = False)

File: src/Json.Schema.ToDotNet/SyntaxNodeExtensions.cs (hunk 2 above)
Comment (michaelcfanning) [reply src/Json.Schema.ToDotNet/SyntaxNodeExtensions.cs]: TIL

---
In reply to: [244507593](https://github.com/Microsoft/jschema/pull/62#discussion_r244507593) [](ancestors = 244507593)

File: src/Json.Schema.ToDotNet/ClassGenerator.cs [hunk 3]
```diff
@@ -30,6 +29,12 @@ public class ClassGenerator : ClassOrInterfaceGenerator
         private const string DefaultValueAttributeNamespaceName = "System.ComponentModel";
         private const string DefaultValueAttributeName = "DefaultValue";
 
+        private const string JsonPropertyAttributeNamespaceName = "Newtonsoft.Json";
+        private const string JsonPropertyAttributeName = "JsonProperty";
+        private const string DefaultValueHandlingPropertyName = "DefaultValueHandling";
+        private const string DefaultValueHandlingEnumerationName = "DefaultValueHandling";
```
[the remark this answers is not in the document]
Comment (lgolding) [reply src/Json.Schema.ToDotNet/ClassGenerator.cs]: Yes, just me being a stickler. The property name "just happens" to be the same as the enumeration name.

---
In reply to: [244508011](https://github.com/Microsoft/jschema/pull/62#discussion_r244508011) [](ancestors = 244508011)

File: src/Json.Schema.ToDotNet/SyntaxNodeExtensions.cs (hunk 2 above)
Comment (lgolding) [reply src/Json.Schema.ToDotNet/SyntaxNodeExtensions.cs]: Yeah, I learned that from one of the IDE diagnostics in the error window and read about it yesterday.

---
In reply to: [244508046](https://github.com/Microsoft/jschema/pull/62#discussion_r244508046) [](ancestors = 244508046,244507593)

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs [hunk 4]
```diff
@@ -1070,6 +1071,7 @@ public SNodeKind SNodeKind
         /// </summary>
         [DataMember(Name = ""intPropWithDefault"", IsRequired = false, EmitDefaultValue = false)]
         [DefaultValue(42)]
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.IgnoreAndPopulate)]
```
Comment (michaelcfanning) [inline src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: >DefaultValueHandling [](start = 22, length = 20)

What's your process for maintaining these tests? Is it only manually? Do you use debugger values for comparison? Do you ever write them to disk and diff? Etc.

Signed, 'Harps on a Consistent Set of Strings'

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs [hunk 5]
```diff
@@ -2363,7 +2369,152 @@ public int GetHashCode(C obj)
...
+        }
+
+        [Fact(DisplayName = "DataModelGenerator generates attributes for properties with defaults.")]
+        public void GeneratesAttributesForPropertiesWithDefaults()
+        {
+            const string Schema =
+@"{
+  ""type"": ""object"",
+  ""properties"": {
+    ""intProp"": {
```
Comment (michaelcfanning) [inline src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: >intProp [](start = 6, length = 7)

why so stingy with your identifiers? Is there anything wrong with 'integerProperty' here? readability of these string literals honestly would be improved. the double-quotes work against the terseness of your ids. fine as a TBD, of course, if you even decide to take the note

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs [hunk 6]
```diff
@@ -2363,7 +2369,152 @@ public int GetHashCode(C obj)
...
+      ""default"": 42.1
+    },
+    ""stringProp"": {
+      ""type"": ""string"",
+      ""description"": ""A string property.""
+    },
+    ""stringPropWithDefault"": {
+      ""type"": ""string"",
+      ""description"": ""A string property with a default value."",
+      ""default"": ""42""
```
Comment (michaelcfanning) [inline src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: >42 [](start = 21, length = 2)

you couldn't think up a better string value here? how about "thanks for all the fish"

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs (hunk 4 above)
Comment (lgolding) [reply src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: I "maintain" it by:

1. Doing a "Copy all" of the error message from the Tests Explorer window.
2. Copying it into a blank file.
3. Splitting the "expected" and "actual" strings into separate files.
4. Using Beyond Compare to convince myself that the new "actual" is correct.
5. Updating the test by hand.

Not fun.

---
In reply to: [244508100](https://github.com/Microsoft/jschema/pull/62#discussion_r244508100) [](ancestors = 244508100)

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs [hunk 7]
```diff
@@ -2363,7 +2369,152 @@ public int GetHashCode(C obj)
...
+  ""type"": ""object"",
+  ""properties"": {
+    ""intProp"": {
+      ""type"": ""integer"",
+      ""description"": ""An integer property.""
+    },
+    ""intPropWithDefault"": {
+      ""type"": ""integer"",
+      ""description"": ""An integer property with a default value."",
+      ""default"": 42
```
Comment (michaelcfanning) [inline src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: >42 [](start = 19, length = 2)

what would happen, I wonder, if you set this value to something that exceeds Int32.MaxValue? Just curious about current symptoms in this circumstance

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs (hunk 6 above)
Comment (lgolding) [reply src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: Oooh! I'm going to change that.

---
In reply to: [244508189](https://github.com/Microsoft/jschema/pull/62#discussion_r244508189) [](ancestors = 244508189)

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs [hunk 8]
```diff
@@ -2363,7 +2369,152 @@ public int GetHashCode(C obj)
...
+    },
+    ""boolPropWithTrueDefault"": {
+      ""type"": ""boolean"",
+      ""description"": ""A Boolean property with a true default value."",
+      ""default"": true
+    },
+    ""boolPropWithFalseDefault"": {
+      ""type"": ""boolean"",
+      ""description"": ""A Boolean property with a false default value."",
+      ""default"": false
```
Comment (michaelcfanning) [inline src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: >false [](start = 19, length = 5)

not sure how interesting this is, since it's the .NET default. why did you include it? if this test is interesting, why not add a string property with the default of 'null'? and an integer with 0, etc? just to be comprehensive.

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs [hunk 9]
```diff
@@ -2363,7 +2369,152 @@ public int GetHashCode(C obj)
...
+        [DataMember(Name = ""boolPropWithTrueDefault"", IsRequired = false, EmitDefaultValue = false)]
+        [DefaultValue(true)]
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.IgnoreAndPopulate)]
+        public bool BoolPropWithTrueDefault { get; set; }
+
+        /// <summary>
+        /// A Boolean property with a false default value.
+        /// </summary>
+        [DataMember(Name = ""boolPropWithFalseDefault"", IsRequired = false, EmitDefaultValue = false)]
+        [DefaultValue(false)]
```
Comment (michaelcfanning) [inline src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: this actually an entirely unnecessary attribution, I believe, as is the JsonProperty attribute that values. For cases where .NET init saves the day, EmitDefaultValue is sufficient. whether you want to handle this case is up to you. Note that there are .NET helpers for retrieving default values of object type, I believe, which might help you create conditional logic to avoid this code emit

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs (hunk 8 above)
Comment (lgolding) [reply src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: It's only there because it exercises a code branch. In Roslyn the `true` literal is a different `SyntaxKind` from the `false` literal, so there's a ternary operator in the code to generate one or the other. There's no similar branch for numeric values. And you can't put `null` for a `string`-valued property; `null` is a different data type.

---
In reply to: [244508342](https://github.com/Microsoft/jschema/pull/62#discussion_r244508342) [](ancestors = 244508342)

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs [hunk 10]
```diff
@@ -2363,7 +2369,152 @@ public int GetHashCode(C obj)
...
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.IgnoreAndPopulate)]
+        public bool BoolPropWithFalseDefault { get; set; }
+    }
+}";
+
+            var generator = new DataModelGenerator(_settings, _testFileSystem.FileSystem);
+            JsonSchema schema = SchemaReader.ReadSchema(Schema, TestUtil.TestFilePath);
```
Comment (michaelcfanning) [inline src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: >TestUtil [](start = 64, length = 8)

'TestUtil', boo. are we programming in the 80's? you can take the engineers out of the eighties i suppose, but you can't completely take the eighties from the engineer. cf. earlier references to 'hitchhiker's guide to the galaxy'. actually published in 10/79, of course, but read by us, no doubt, in the subsequent decade

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs (hunk 5 above)
Comment (lgolding) [reply src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: I don't mind changing them. Doing it now...

---
In reply to: [244508136](https://github.com/Microsoft/jschema/pull/62#discussion_r244508136) [](ancestors = 244508136)

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs [hunk 11]
```diff
@@ -2363,7 +2369,152 @@ public int GetHashCode(C obj)
...
+
+            generator.Generate(schema);
+
+            var expectedContentsDictionary = new Dictionary<string, ExpectedContents>
```
Comment (michaelcfanning) [inline src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: >expectedContentsDictionary [](start = 16, length = 26)

your fancy initialization here is working against readabilty, for those not familiar with the indexing syntax you're utilizing.

Here's another way to do it:

            var expectedContentsDictionary = new Dictionary<string, ExpectedContents>
            {
                { _settings.RootClassName,  new ExpectedContents {ClassContents = ExpectedClass } }
            };

Even putting your current code on one line helps communicate what's happening better, I think.

            var expectedContentsDictionary = new Dictionary<string, ExpectedContents>
            {
                [_settings.RootClassName] = new ExpectedContents { ClassContents = ExpectedClass }
            };

summary of note: improve readability if you can

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs [hunk 12]
```diff
@@ -3464,5 +3615,10 @@ public sealed class C
             string actual = generator.Generate(schema);
             actual.Should().Be(Expected);
         }
+
+        private void VerifyGeneratedFileContents(IDictionary<string, ExpectedContents> expectedContentsDictionary)
```
Comment (michaelcfanning) [inline src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: >VerifyGeneratedFileContents [](start = 21, length = 27)

why did you do this? I am suspicious of helpers that consist of  a single line of code. do you project more shared validation here in future?

the downside of what you've done, taken to an extreme, is endless stepping into/out during debugging sessions. as you know, i approach unit-testing to maximize debuggability/ease of update.

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs (hunk 4 above)
Comment (michaelcfanning) [reply src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: Dear god in heaven. There are several (unfortunately diverse) techniques demonstrated in the SARIF-SDK that would alleviate this. The best is to send all output to dedicated 'expected' and 'actual' directories. Then you can diff the dir for a large suite of tests. You can copy 'actual' files into 'expected' once you confirm that you like them. 

---
In reply to: [244508254](https://github.com/Microsoft/jschema/pull/62#discussion_r244508254) [](ancestors = 244508254,244508100)

File: src/Json.Schema.ToDotNet.UnitTests/Hints/DictionaryHintTests.cs [hunk 13]
```diff
@@ -1245,7 +1245,7 @@ public partial class Def2
                 }
             };
 
-            Assert.FileContentsMatchExpectedContents(TestFileSystem, expectedContentsDictionary);
+            Assert.FileContentsMatchExpectedContents(TestFileSystem, expectedContentsDictionary, Settings.GenerateEqualityComparers);
```
Comment (michaelcfanning) [inline src/Json.Schema.ToDotNet.UnitTests/Hints/DictionaryHintTests.cs]: >Settings [](start = 97, length = 8)

'Settings' is too generic a term here.

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs (hunk 5 above)
Comment (michaelcfanning) [reply src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: ugh. sorry to put you through this. wish you had an automated process. :) i can't resist asking, how do you handle the double quote escaping? do you have any file replace magic for this? 

---
In reply to: [244508723](https://github.com/Microsoft/jschema/pull/62#discussion_r244508723) [](ancestors = 244508723,244508136)

Comment (lgolding) [reply src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: No, I just type the double-quotes. Anyway, this one is now fixed.

---
In reply to: [244510035](https://github.com/Microsoft/jschema/pull/62#discussion_r244510035) [](ancestors = 244510035,244508723,244508136)

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs (hunk 12 above)
Comment (lgolding) [reply src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: Once I added the third argument, and two of the three arguments were instance fields, I felt this was more compact.

---
In reply to: [244509301](https://github.com/Microsoft/jschema/pull/62#discussion_r244509301) [](ancestors = 244509301)

**Comment by michaelcfanning** [comment]
>         protected override AttributeSyntax[] GeneratePropertyAttributes(string propertyName, string serializedName, bool isRequired, object defaultValue)

agreed, and fine to defer. I've approved this change which in the main is just fine. I gave you the full bore review, of course, since it's the weekend and I have the luxury of focusing on it. :) I will continue to track your changes/any discussion but fire away when you are satisfied.

---
In reply to: [450516997](https://github.com/Microsoft/jschema/pull/62#issuecomment-450516997) [](ancestors = 450516997)

---
Refers to: src/Json.Schema.ToDotNet/ClassGenerator.cs:252 in ad3e739. [](commit_id = ad3e7397e99d050f0a65872564a2063bb1d8f97d, deletion_comment = False)

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs (hunk 11 above)
Comment (lgolding) [reply src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: Putting it on one line is ok if there's only one initializer, but it spreads out too far when there is more than one:
```
            var expectedContentsDictionary = new Dictionary<string, ExpectedContents>
            {
                [_settings.RootClassName] = new ExpectedContents
                {
                    ClassContents = ExpectedRootClass,
                    ComparerClassContents = ExpectedRootComparerClass
                },
                ["Color"] = new ExpectedContents
                {
                    ClassContents = ExpectedColorClass,
                    ComparerClassContents = ExpectedColorComparerClass
                }
            };
```
IMO this is fine, and I'm not sympathetic to "hard to read if you don't know the language" ;-)

---
In reply to: [244509144](https://github.com/Microsoft/jschema/pull/62#discussion_r244509144) [](ancestors = 244509144)

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs (hunk 10 above)
Comment (lgolding) [reply src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: Ugh. I'm not changing the 53 occurrences of this name. We can laugh at our younger selves, but I don't see this name as a barrier to readability. It's just a bit old-fashioned. That's why Terisa calls me "Fifties Man".

---
In reply to: [244508719](https://github.com/Microsoft/jschema/pull/62#discussion_r244508719) [](ancestors = 244508719)

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs (hunk 9 above)
Comment (lgolding) [reply src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: Nice idea. I filed [#63](https://github.com/Microsoft/jschema/issues/63), "Don't emit default-related attributes if the JSON schema default is the same as the .NET default", and will address it separately (but not with high priority -- it was important to get this far, for the sake of the SARIF schema changes, but we can certainly live without #63).

I found [this helper](https://stackoverflow.com/questions/2490244/default-value-of-a-type-at-runtime/2490274#2490274):
```
object GetDefaultValue(Type t)
{
    if (t.IsValueType)
        return Activator.CreateInstance(t);

    return null;
}
```



---
In reply to: [244508499](https://github.com/Microsoft/jschema/pull/62#discussion_r244508499) [](ancestors = 244508499)

**Comment by lgolding** [comment]
@mikefan@microsoft.com Now that this is merged, can you update the version number and publish the new package?

**Comment by michaelcfanning** [comment]
Yes.

---
In reply to: [450522566](https://github.com/Microsoft/jschema/pull/62#issuecomment-450522566) [](ancestors = 450522566)

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs (hunk 10 above)
Comment (michaelcfanning) [reply src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: Born in the Fifties. Formed by the Eighties. Perhaps you enjoyed the Police song 'Born in the Fifties' at that time.

---
In reply to: [244512315](https://github.com/Microsoft/jschema/pull/62#discussion_r244512315) [](ancestors = 244512315,244508719)

File: src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs (hunk 11 above)
Comment (michaelcfanning) [reply src/Json.Schema.ToDotNet.UnitTests/DataModelGeneratorTests.cs]: Code is written to serve the reader and not the author. Patterns at the leading edge of the language are often confusing to people. There often isn't common convention around how best to expresss them, as various codesmiths are tinkering. My note here is only that this fairly abstruse initialization syntax is not particularly readable. 

---
In reply to: [244511949](https://github.com/Microsoft/jschema/pull/62#discussion_r244511949) [](ancestors = 244511949,244509144)
